src/services/jsTyping.ts

```diff
@@ -5,58 +5,205 @@
+        : { cachedTypingPaths: string[], newTypingNames: string[] } {
+        _host = host;
+        // Directories to search for package.json, bower.json and other typing information
         let searchDirs: string[] = [];
-        let typingNames: string[] = [];
-        let cachedTypingPaths: string[] = [];
-        let cachedTypingNames: string[] = [];
 
         for (let fileName of fileNames) {
-            let dir = ts.getDirectoryPath(ts.normalizePath(fileName));
-            if (searchDirs.indexOf(dir) < 0) {
-                searchDirs.push(dir);
+            if (ts.getBaseFileName(fileName) !== "lib.d.ts") {
+                let dir = ts.getDirectoryPath(ts.normalizePath(fileName));
+                if (searchDirs.indexOf(dir) < 0) {
+                    searchDirs.push(dir);
+                }
             }
         }
 
         for (let searchDir of searchDirs) {
             let packageJsonPath = ts.combinePaths(searchDir, "package.json");
-            if (host.fileExists(packageJsonPath)) {
-                let packageJsonDict = JSON.parse(host.readFile(packageJsonPath));
-                if (packageJsonDict.hasOwnProperty("dependencies")) {
-                    typingNames = typingNames.concat(Object.keys(packageJsonDict.dependencies));
-                }
-            }
+            getTypingNamesFromJson(packageJsonPath);
+
+            let bowerJsonPath = ts.combinePaths(searchDir, "bower.json");
+            getTypingNamesFromJson(bowerJsonPath);
+
+            let nodeModulesPath = ts.combinePaths(searchDir, "node_modules");
+            getTypingNamesFromNodeModuleFolder(nodeModulesPath);
         }
         
+        // Todo: use a real safe list
+        getTypingNamesFromSourceFileNames(fileNames, ["react", "jquery"]);
+        getTypingNamesFromCompilerOptions(compilerOptions);
+
         let normalizedCachePath = ts.normalizePath(cachePath);
         let typingsPath = ts.combinePaths(normalizedCachePath, "typings");
         let tsdJsonCacheInfo = tryGetTsdJsonCacheInfo(host, cachePath);
         if (tsdJsonCacheInfo) {
-            let cacheTsdJsonDict = tsdJsonCacheInfo.value;
-            if (cacheTsdJsonDict.hasOwnProperty("notFound")) {
-                let notFound = cacheTsdJsonDict["notFound"];
-                typingNames = ts.filter(typingNames, i => notFound.indexOf(i) < 0);
-            }
-            if (cacheTsdJsonDict.hasOwnProperty("installed")) {
-                for (let cachedTypingPath of Object.keys(cacheTsdJsonDict.installed)) {
-                    let cachedTypingName = cachedTypingPath.substr(0, cachedTypingPath.indexOf('/'));
-                    if (typingNames.indexOf(cachedTypingName) >= 0) {
-                        cachedTypingNames.push(cachedTypingName);
-                        cachedTypingPaths.push(ts.combinePaths(typingsPath, cachedTypingPath));
+            try {

```

Comment:
do we need this try-catch?  It looks like we shouldn't be throwing here if we check:
if (tsdJsonCacheInfo && tsdJsonCacheInfo.value)